pkg/controller/egress/controller_test.go

```diff
@@ -136,6 +138,51 @@ type egressController struct {
 func newController(objects, crdObjects []runtime.Object) *egressController {
 	client := fake.NewSimpleClientset(objects...)
 	crdClient := fakeversioned.NewSimpleClientset(crdObjects...)
+	// These reactors are in charge of incrementing Generation for Egress resources, so that
+	// changes done through crdClient are not ignored by the EgressController.
+	// Unfortunately, it seems that we have no choice but accessing the ObjectTracker directly.
```

Comment:
Why increasing the generation of the object in the action doesn't work? I remember it worked when I was fixing another issue: https://github.com/kubernetes/kubernetes/commit/c2d3e54551e5cd84660b212b9f5bea2fdb64a43a#diff-7ee3ee72ee837df58329263f1979616a693a64223fb16c7e9e88a5974a16e9bfR410.
The last reactor is responsible for storing the object in ObjectTracker, so as long as the change we made in a prepended reactor should take effect.
https://github.com/antrea-io/antrea/blob/3a51abe9562aad96a8bd486d6fd90f679c8cced6/pkg/client/clientset/versioned/fake/clientset_generated.go#L56

Reply:
* for patch action: the action doesn't actually include the object (unlike update), so there is no access to the Generation field. See https://pkg.go.dev/k8s.io/client-go/testing#PatchAction. Any object returned by the reactor (second return value of the reactor) will also be ignored by `ObjectReaction` when using patch.
* but even for update action it will not work here. I originally had a simpler reactor for update, and the test was still failing sometimes (with a different error). This is the reactor I had:
```
egressUpdateReactor := k8stesting.ReactionFunc(
		func(action k8stesting.Action) (handled bool, ret runtime.Object, err error) {
			ua := action.(k8stesting.UpdateAction)
			egress := ua.GetObject().(*v1alpha2.Egress)
			egress.Generation = egress.Generation + 1
			return false, egress, nil
		})
```
This is also incorrect because the Egress resource passed to Egresses().Update may have an invalid or stale Generation value, and that stale value will end up being incremented and stored in the ObjectTracker. So it is subject to races between the unit test code and the controller code.

Reply:
BTW, I considered using `egress.Generation = rand()`, which would simplify things a little. But we would still need to access the ObjectTracker directly for the "patch" case. Using incremental Generation number is also less surprising and easier to debug IMO.

Reply:
Thanks for the explanation. Do you think if a closure could work here?
```
	egressUpdateReactor, egressPatchReactor := func() (k8stesting.ReactionFunc, k8stesting.ReactionFunc) {
		generation := int64(0)
		updateReactor := func(action k8stesting.Action) (handled bool, ret runtime.Object, err error) {
			ua := action.(k8stesting.UpdateAction)
			egress := ua.GetObject().(*v1alpha2.Egress)
			generation += 1
			egress.Generation = generation
			return false, egress, nil
		}
		patchReactor := func(action k8stesting.Action) (handled bool, ret runtime.Object, err error) {
			ua := action.(k8stesting.PatchActionImpl)
			patch := map[string]interface{}{}
			json.Unmarshal(ua.GetPatch(), &patch)
			generation += 1
			patch["generation"] = generation
			ua.Patch, _ = json.Marshal(patch)
			return false, nil, nil
		}
		return updateReactor, patchReactor
	}()
```

Reply:
It should work. I can also make it a map so that 2 Egress resources don't share the same generation counter. I will run the tests again with the new version.

Reply:
Updated, thanks for the suggestion. I made a few changes to your proposal and added some comments, but it's the same idea

Reply:
I had to find another workaround, because `action` cannot be mutated in the patch reactor (it's an object and not a pointer to an object): when testing in Github CI, it failed because generation was not patched correctly.
The new solution is getting a bit complicated but functionally it's probably still better than my original one because we avoid multiple updates to the ObjectTracker for patch.